javalib/src/main/scala/java/lang/String.scala

```diff
@@ -1,42 +1,908 @@
 package java.lang
 
-import java.io.Serializable
-
-class _String private() extends CharSequence with Serializable with Comparable[_String] {
-  def this(value: Array[scala.Char], offset: scala.Int, count: scala.Int) = this()
-
-  def length(): scala.Int = ???
-  def charAt(index: scala.Int): Char = ???
-  def compareTo(another_String: _String): scala.Int = ???
-  def subSequence(beginIndex: scala.Int, endIndex: scala.Int): CharSequence = ???
-  def concat(str: _String): _String = ???
-  def indexOf(ch: scala.Int): scala.Int = ???
-  def indexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def substring(beginIndex: scala.Int): _String = ???
-  def substring(beginIndex: scala.Int, endIndex: scala.Int): _String = ???
-  def startsWith(prefix: _String): scala.Boolean = ???
-  def toLowerCase(): _String = ???
-  def toCharArray(): Array[Char] = ???
-  def equalsIgnoreCase(str: _String): scala.Boolean = ???
-  def getChars(srcBegin: scala.Int, srcEnd: scala.Int, dst: Array[scala.Char],
-      dstBegin: scala.Int): Unit = ???
-  override def hashCode(): scala.Int = ???
-
-  // TODO: rest of the api
+import java.io.{Serializable, UnsupportedEncodingException}
+import java.util._
+import java.util.regex._
+import java.nio._
+import java.nio.charset._
+
+import scala.util.Try
+
+final class _String private ()
+    extends Serializable with Comparable[_String] with CharSequence {
+  private var value         = new Array[Char](0)
+  private var offset        = 0
+  private var count         = 0
+  private var hashcode: Int = _
+
+  def this(data: Array[scala.Byte], start: Int, length: Int) = {
+    this()
+    if (start >= 0 && 0 <= length && length <= data.length - start) {
+      offset = 0
+      val charset    = defaultCharset()
+      val charBuffer = charset.decode(ByteBuffer.wrap(data, start, length))
+      if (charBuffer.length > 0) {
+        value = charBuffer.array
+        count = charBuffer.length
+      } else {
+        count = 0
+        value = new Array[Char](0)
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  def this(data: Array[scala.Byte], _high: Int, start: Int, length: Int) = {
+    this()
+    if (length <= data.length - start && start >= 0 && 0 <= length) {
+      offset = 0
+      value = new Array[Char](length)
+      var high = _high
+      high <<= 8
+      for (i <- 0 until length) {
+        value(i) = (high + data(start + i) & 0xff).toChar
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  def this(
+      data: Array[scala.Byte], start: Int, length: Int, encoding: _String) = {
+    this()
```

Comment:
This constructor can be written very simply as `this(data, start, length, Charset.forName(encoding))`